multicluster traffic management (#10486)
istio/istio.io: content/en/docs/ops/configuration/traffic-management/multicluster/index.md

---
title: Multi-cluster Traffic Management
description: How to configure how traffic is distributed among clusters in the mesh.
weight: 70
keywords: [traffic-management,multicluster]
owner: istio/wg-networking-maintainers
test: no
---

> Within a multicluster mesh, [namespace sameness](https://github.com/kubernetes/community/blob/master/sig-multicluster/namespace-sameness-position-statement.md)

Author: Is this the right place for the sameness note? Or should it live in `deployment models`? Or both?

Reviewer: This should definitely go in deployment models. Maybe we need a glossary entry as well for `namespace sameness`. I'm fine duplicating the link here.

applies and all namespaces with a given name are considered to be the same namespace. If multiple clusters contain a
`Service` with the same namespaced name, they will be recognized as a single combined service. By default, traffic is
load-balanced across all clusters in the mesh for a given service. In some cases, that behavior is not desirable, and
this document describes the different options for overriding that default.

> ## Cluster Local Services

Author: should we also have a section that highlights locality load balancing as an option? and is this header correct?

### `DestinationRule` subsets 

Using the label `topology.istio.io/cluster` in the subset selector for a DestinationRule, you can create
per-cluster subsets.

{{< text yaml >}}
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: mysvc-dr
spec:
  host: mysvc.myns.svc.cluster.local
  subsets:
  - name: cluster-1
    labels:
      topology.istio.io/cluster: cluster-1
  - name: cluster-2
    labels:
      topology.istio.io/cluster: cluster-2
{{< /text >}}

To restrict traffic to only the local cluster, add a VirtualService. You can extend this configuration
to apply traffic shifting or mirroring rules on a per-cluster basis.

{{< text yaml >}}
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
name: mysvc-vs
spec:
  hosts:
  - mysvc.myns.svc.cluster.local

>   http:

Author: without `sourceLabels` support the VS has to be different in each cluster...

Author: https://github.com/istio/istio/pull/35998

It actually already works :)

  - name: "cluster-1-local"
    route:
    - destination:
      host: mysvc.myns.svc.cluster.local
      subset: cluster-1
{{< /text >}}

### Per-cluster `Service` or `Namespace`

> If `Services` aren't meant to be used cross-cluster at all, it may make sense to simply give them unique names in each

Reviewer: Can we come up with a more concrete example for this? I'm having trouble picturing how this is different from cluster-local.

---
title: Multi-cluster Traffic Management
description: How to configure how traffic is distributed among clusters in the mesh.
weight: 70
keywords: [traffic-management,multicluster]
owner: istio/wg-networking-maintainers
test: no
---

### Services that are not the same

Within a multicluster mesh, [namespace sameness](https://github.com/kubernetes/community/blob/master/sig-multicluster/namespace-sameness-position-statement.md)
applies and all namespaces with a given name are considered to be the same namespace. If multiple clusters contain a
`Service` with the same namespaced name, they will be recognized as a single combined service. By default, traffic is
load-balanced across all clusters in the mesh for a given service. 

If there is no case where this traffic should be sent across clusters, consider giving the services a different name in each cluster.

### Partitioning Service Endpoints (i.e. creating subsets) 

> Using the label `topology.istio.io/cluster` in the subset selector for a DestinationRule, you can create

Reviewer 2: +1 I think a list of the available built-in labels that one can use, would be really good.

> Using the label `topology.istio.io/cluster` in the subset selector for a DestinationRule, you can create

Author: Sgtm. Would be cool if we included something in [/reference/config/labels/](https://istio.io/latest/docs/reference/config/labels/) that additionally highlights the which ones are built-in.

> 

Reviewer 2: ```suggestion

For example, you can enforce cluster-local traffic for an individual service, all services in a particular namespace, or globally for all services in the mesh, as follows:

```

{{< tabset category-name="meshconfig" >}}

{{< tab name="per-service" category-value="service" >}}

> {{< text yaml >}}

Reviewer 2: Should give more context. What CRD is this, and were is it applied?

Author: Will do.

serviceSettings:
- settings:
  clusterLocal: true
  hosts:

>     - "mysvc.myns.svc.cluster.local"

Reviewer 2: This should be indented more?

btw, where is the proto for this?

Author: good catch – and I'm not sure what you mean by where is the proto?

  - name: cluster-1
    labels:
      topology.istio.io/cluster: cluster-1
  - name: cluster-2
    labels:
      topology.istio.io/cluster: cluster-2
{{< /text >}}

Using these subsets you can create various routing rules based on the cluster such as [mirroring](/docs/tasks/traffic-management/mirroring/)
or [shifting](/docs/tasks/traffic-management/traffic-shifting/).

### Keeping traffic in-cluster

In some cases the default cross-cluster load balancing behavior is not desirable. To keep traffic "cluster-local" (i.e.

> traffic sent from `cluster-a` will only reach destinations in `cluster-a`), there are multiple-approaches.

Reviewer 3: there is another approach localityloadbalancing.FailoverPriority

Author: Yes! Definitely want to fit that in.

#### Restrict destination subsets in `VirtualService`

{{< text yaml >}}
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: mysvc-vs
spec:
  hosts:
  - mysvc.myns.svc.cluster.local
  http:
  - name: "cluster-1-local"
    match:
    - sourceLabels:
        topology.istio.io/cluster: "cluster-1"
    route:
    - destination:
        host: mysvc.myns.svc.cluster.local
        subset: cluster-1
  - name: "cluster-2-local"
    match:
    - sourceLabels:
        topology.istio.io/cluster: "cluster-2"
    route:
    - destination:
        host: mysvc.myns.svc.cluster.local
        subset: cluster-2
{{< /text >}}

> #### Global `MeshConfig` settings

Reviewer 2: That sounds good to me. It would be good to also explain that service partitioning can also be used to keep traffic in-cluster and maybe touch on why use one vs the other.

> #### Global `MeshConfig` settings

Author: I've put things in this structure, but I can't really think of a good way to word the pros/cons of each of the "cluster-local" approaches. 

On one hand, the subsetting approach allows a bit more granular control, but it also starts to mix service-level policy with topology-level policy. For example, a config that sends 10% of traffic to `v2` now needs 2x subsets (i.e. `cluster-1-v2`, `cluster-2-v2`)

> #### Global `MeshConfig` settings

Reviewer 2: Rather than pros/cons I was thinking just a high level statement, maybe along the lines of what you just said. The one is generally more appropriate for setting topology-level policy, while the other is more service focused.

> #### Global `MeshConfig` settings

Author: added a section at the bottom

    labels:
      topology.istio.io/cluster: cluster-1
  - name: cluster-2
    labels:
      topology.istio.io/cluster: cluster-2
{{< /text >}}

Using these subsets you can create various routing rules based on the cluster such as [mirroring](/docs/tasks/traffic-management/mirroring/)
or [shifting](/docs/tasks/traffic-management/traffic-shifting/).

This provides another option to create cluster-local traffic rules by restricting the destination subset in a `VirtualService`:

{{< text yaml >}}
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: mysvc-cluster-local-vs
spec:
  hosts:
  - mysvc.myns.svc.cluster.local
  http:
  - name: "cluster-1-local"
    match:
    - sourceLabels:
        topology.istio.io/cluster: "cluster-1"
    route:
    - destination:
        host: mysvc.myns.svc.cluster.local
        subset: cluster-1
  - name: "cluster-2-local"
    match:
    - sourceLabels:
        topology.istio.io/cluster: "cluster-2"
    route:
    - destination:
        host: mysvc.myns.svc.cluster.local
        subset: cluster-2
{{< /text >}}

> The subset approach can accomplish the same goals as [`MeshConfig.serviceSettings`](/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig-ServiceSettings-Settings),
> but mixes service-level policy with topology-level policy. For example, a rule that sends 10% of traffic to `v2` of some service now needs twice the
> number of subsets (i.e. `cluster-1-v2`, `cluster-2-v2`). This approach should be preferred only when you need more granular control of when to route
> based on the cluster.

Reviewer 2: ```suggestion
Using subset-based routing this way to control cluster-local traffic, as opposed to
[`MeshConfig.serviceSettings`](/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig-ServiceSettings-Settings),
has the downside of mixing service-level policy with topology-level policy.
For example, a rule that sends 10% of traffic to `v2` of a service will need twice the
number of subsets (e.g., `cluster-1-v2`, `cluster-2-v2`).
This approach is best limited to situations where more granular control of cluster-based routing is needed.
```

1. Make sure your deployed services follow the concept of {{< gloss "namespace sameness" >}}namespace sameness{{< /gloss >}}.

## Keeping traffic in-cluster

In some cases the default cross-cluster load balancing behavior is not desirable. To keep traffic "cluster-local" (i.e.
traffic sent from `cluster-a` will only reach destinations in `cluster-a`), mark hostnames or wildcards as `clusterLocal`
using [`MeshConfig.serviceSettings`](/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig-ServiceSettings-Settings).

For example, you can enforce cluster-local traffic for an individual service, all services in a particular namespace, or globally for all services in the mesh, as follows:

{{< tabset category-name="meshconfig" >}}

{{< tab name="per-service" category-value="service" >}}

{{< text yaml >}}
serviceSettings:
- settings:
  clusterLocal: true
  hosts:
    - "mysvc.myns.svc.cluster.local"
{{< /text >}}

{{< /tab >}}

{{< tab name="per-namespace" category-value="namespace" >}}

{{< text yaml >}}
serviceSettings:
- settings:
  clusterLocal: true
  hosts:
    - "*.myns.svc.cluster.local"
{{< /text >}}

{{< /tab >}}

> {{< tab name="global" category-value="service" >}}

Reviewer 2: Tabs aren't displaying properly in preview because of this:
```suggestion
{{< tab name="global" category-value="global" >}}
```

keywords: [traffic-management,multicluster]
owner: istio/wg-networking-maintainers
test: no
---

Within a multicluster mesh, traffic rules specific to the cluster topology may be desirable. This document describes
a few ways to manage traffic in a multicluster mesh. Before reading this guide:

1. Read [Deployment Models](/docs/ops/deployment/deployment-models/#multiple-clusters)
1. Make sure your deployed services follow the concept of {{< gloss "namespace sameness" >}}namespace sameness{{< /gloss >}}.

## Keeping traffic in-cluster

In some cases the default cross-cluster load balancing behavior is not desirable. To keep traffic "cluster-local" (i.e.
traffic sent from `cluster-a` will only reach destinations in `cluster-a`), mark hostnames or wildcards as `clusterLocal`
using [`MeshConfig.serviceSettings`](/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig-ServiceSettings-Settings).

For example, you can enforce cluster-local traffic for an individual service, all services in a particular namespace, or globally for all services in the mesh, as follows:

{{< tabset category-name="meshconfig" >}}

{{< tab name="per-service" category-value="service" >}}

{{< text yaml >}}

> serviceSettings:
> - settings:
>   clusterLocal: true
>   hosts:
>     - "mysvc.myns.svc.cluster.local"
> {{< /text >}}
> 
> {{< /tab >}}
> 
> {{< tab name="per-namespace" category-value="namespace" >}}
> 
> {{< text yaml >}}
> serviceSettings:
> - settings:
>   clusterLocal: true
>   hosts:
>     - "*.myns.svc.cluster.local"
> {{< /text >}}
> 
> {{< /tab >}}
> 
> {{< tab name="global" category-value="global" >}}
> 
> {{< text yaml >}}
> serviceSettings:
> - settings:
>   clusterLocal: true
>   hosts:
>     - "*"

Reviewer 2: Oops, format still off here.
```suggestion
serviceSettings:
- settings:
    clusterLocal: true
  hosts:
  - "mysvc.myns.svc.cluster.local"
{{< /text >}}

{{< /tab >}}

{{< tab name="per-namespace" category-value="namespace" >}}

{{< text yaml >}}
serviceSettings:
- settings:
    clusterLocal: true
  hosts:
  - "*.myns.svc.cluster.local"
{{< /text >}}

{{< /tab >}}

{{< tab name="global" category-value="global" >}}

{{< text yaml >}}
serviceSettings:
- settings:
    clusterLocal: true
  hosts:
  - "*"
```

kind: VirtualService
metadata:
  name: mysvc-cluster-local-vs
spec:
  hosts:
  - mysvc.myns.svc.cluster.local
  http:
  - name: "cluster-1-local"
    match:
    - sourceLabels:
        topology.istio.io/cluster: "cluster-1"
    route:
    - destination:
        host: mysvc.myns.svc.cluster.local
        subset: cluster-1
  - name: "cluster-2-local"
    match:
    - sourceLabels:
        topology.istio.io/cluster: "cluster-2"
    route:
    - destination:
        host: mysvc.myns.svc.cluster.local
        subset: cluster-2
{{< /text >}}

### Locality Load Balancing

When using a multi-cluster mesh for redundancy and resiliency, it may be desirable to only send cross-cluster traffic
when necessary. Using [Locality Load Balancing](/docs/tasks/traffic-management/locality-load-balancing/) settings, you
can configure locality weighted distribution or failover.

On top of default locality components of `region/zone/subzone`, you can use [`failoverPriority`](/docs/reference/config/networking/destination-rule/#LocalityLoadBalancerSetting)
to failover based on the cluster or network explicitly. Using the following destination rule, traffic will prefer first

> in-cluster traffic, then in-network, then in-region, etc.

Reviewer 3: This desc is not quite right

{{< text yaml >}}
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: mysvc-cluster-failover
spec:
  host: mysvc.myns.svc.cluster.local
  trafficPolicy:
    loadBalancer:
      simple: ROUND_ROBIN
      localityLbSetting:
        enabled: true
        failoverPriority:
        - "topology.istio.io/cluster"
        - "topology.istio.io/network"
        - "topology.kubernetes.io/region"
        - "topology.kubernetes.io/zone"

>         - "topology.istio.io/subzone"

Reviewer 3: maybe you can remove the region/zone/subzone

If there is no case where this traffic should be sent across clusters, consider giving the services a different name in each cluster.

### Locality Load Balancing

When using a multi-cluster mesh for redundancy and resiliency, it may be desirable to only send cross-cluster traffic
when necessary. Using [Locality Load Balancing](/docs/tasks/traffic-management/locality-load-balancing/) settings, you
can configure locality weighted distribution or failover.

On top of default locality components of `region/zone/subzone`, you can use [`failoverPriority`](/docs/reference/config/networking/destination-rule/#LocalityLoadBalancerSetting)
to failover based on the cluster or network explicitly. Using the following destination rule, traffic will prefer first

> in-cluster traffic, then in-network, then in-region, etc.

Reviewer 3: Not exactly, matching all the labels first, then second, match the first N-1 labels.
